Add Quaternion FunctionOfTime

knelli2 opened the pull request:
## Proposed changes

Adds a new `FunctionOfTime` that stores quaternions which will be used in the time-dependent rotation maps. To get the quaternion, an ODE is solved of the form `dtquat = 0.5 * quat x omega` where `omega` is the orbital angular velocity which is retreived from a `PiecewisePolynomial` in the global cache, and `x` here is quaternion multiplication. Different from a `PiecewisePolynomial`, only the quaternion itself is stored, not any of the derivatives because the derivatives just be calculated from the solved ODE at every function call. Also because derivatives are not stored, what would be the `MaxDeriv` cannot be updated. Before any function call, the `QuaternionFunctionOfTime` checks if it has the same number of stored quaternions as the angular velocity `PiecewisePolynomial` has stored omegas. This is to ensure we have all the data necessary to solve the ODE at any time `t`. Other than these few, but important, differences, the `QuaternionFuntionOfTime` is essentially identical to the `PiecewisePolynomial`.

Also had to add another member to `PiecewisePolynomial` that returns its stored derivative info because `QuaternionFunctionOfTime` needs access to this to solve the ODE so it can populate its missing quaternions at the same times as the stored omegas in the `PiecewisePolynomail`.

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Dependencies

- [x] PR #3310 
- [x] PR #3321 

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

markscheel commented on src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.hpp:
```diff
@@ -0,0 +1,119 @@
+/// the same number of stored quaternions as the angular velocity
+/// `PiecewisePolynomial` has stored omegas. This is to ensure we have all the
+/// data necessary to solve the ODE at any time `t`. Other than these few, but
+/// important, differences, the `QuaternionFuntionOfTime` is essentially
```
[nitpick] spelling `QuaternionFunctionOfTime`

markscheel commented on src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.hpp:
```diff
@@ -0,0 +1,119 @@
+
+  /// Returns the quaternion and the first two derivatives at an arbitrary
+  /// time `t`.
+
```
[nitpick] remove empty line

markscheel commented on src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.hpp:
```diff
@@ -0,0 +1,119 @@
+  double expiration_time_{std::numeric_limits<double>::lowest()};
+
+  /// Solves the ODE \f$ \dot{q} = \frac{1}{2} q \times \omega \f$ at time `t`
+  /// and stores the result in `quaternion_to_integrate`
```
How about "Integrates the ODE \f$ \dot{q} = \frac{1}{2} q \times \omega \f$ from time `t0` to time `t`.  On input, `quaternion_to_integrate` contains the quaternion at time `t0`, and on output, `quaternion_to_integrate` contains the quaternion at time `t`".  (I'm assuming that's where the initial value comes from).

markscheel commented on src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.hpp:
```diff
@@ -0,0 +1,119 @@
+
+  /// Solves the ODE \f$ \dot{q} = \frac{1}{2} q \times \omega \f$ at time `t`
+  /// and stores the result in `quaternion_to_integrate`
+  void solve_quaternion_ode(
+      const double t, const double t0,
+      std::array<DataVector, 1>& quaternion_to_integrate) const noexcept;
```
Our convention in SpECTRE is that non-const arguments that represent output come first in the argument list, and are passed as `gsl::not_null` pointers.
so
```
void solve_quaternion_ode(const gsl::not_null<std::array<DataVector, 1>*> quaternion_to_integrate,
     const double t, const double t0) const noexcept;
```

markscheel commented on src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.hpp:
```diff
@@ -0,0 +1,119 @@
+
+ private:
+  mutable std::vector<StoredInfo<1>> stored_quaternions_and_times_;
+  domain::FunctionsOfTime::PiecewisePolynomial<MaxDerivReturned>*
+      omega_f_of_t_ptr_;
```
This should probably be a `shared_ptr` (or at least something other than a raw pointer).

knelli2 replied:
Tried using `shared_ptr` and ran into errors converting this to/from a `gsl::not_null`. Also not sure the best way to store the omega PiecewisePolynomial inside the QuaternionFunctionOfTime. (See comment below)

markscheel commented on src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.cpp:
```diff
@@ -0,0 +1,259 @@
+  return std::make_unique<QuaternionFunctionOfTime>(*this);
+}
+
+template <size_t MaxDerivReturned>
+void QuaternionFunctionOfTime<MaxDerivReturned>::pup(PUP::er& p) {
+  FunctionOfTime::pup(p);
+  p | stored_quaternions_and_times_;
+  p | expiration_time_;
+}
```
What happens to `omega_f_of_t_ptr_` when this class is pupped?

knelli2 replied:
See comment below.

markscheel commented on src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.cpp:
```diff
@@ -0,0 +1,259 @@
+              std::array<DataVector, 1>>{});
+
+  // Initialize the stepper with initial quaternion (which is the quaternion
+  // passed in), intial time, and intial time step = 1e-3
```
[nitpick] spelling.

markscheel commented on src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.cpp:
```diff
@@ -0,0 +1,259 @@
+  // passed in), intial time, and intial time step = 1e-3
+  dense_stepper.initialize(quaternion_to_integrate, t0, 1e-3);
+
+  // Do an intial step
```
[nitpick] more spelling.

markscheel commented on src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.cpp:
```diff
@@ -0,0 +1,259 @@
+                                   quaternion_to_datavector(dt2quat)};
+}
+
+template <size_t MaxDerivReturned>
+std::array<DataVector, 4>
+QuaternionFunctionOfTime<MaxDerivReturned>::func_and_3_derivs(
+    double t) const noexcept {
```
I think this function should `static_assert` that MaxDerivReturned >=3.  Similar for the other functions that return a fixed number of derivatives.

markscheel commented on tests/Unit/Domain/FunctionsOfTime/Test_QuaternionFunctionOfTime.cpp:
```diff
@@ -0,0 +1,66 @@
+#include "Domain/FunctionsOfTime/PiecewisePolynomial.hpp"
+#include "Domain/FunctionsOfTime/QuaternionFunctionOfTime.hpp"
+
+SPECTRE_TEST_CASE("Unit.Domain.FunctionsOfTime.QuaternionFuntionOfTime",
```
spelling "Funtion"

markscheel commented on tests/Unit/Domain/FunctionsOfTime/Test_QuaternionFunctionOfTime.cpp:
```diff
@@ -0,0 +1,66 @@
+    const std::array<DataVector, 3> quat_func_and_2_derivs =
+        qfot.func_and_2_derivs(check_time);
+
+    // Analytic solution for constant omega
```
Please add a test for non-constant omega as well as this test for constant omega. Perhaps omega linear or quadratic in time. [recall that an early version of the SpEC QuatFvt was correct for constant omega but wrong for non-constant omega].

markscheel commented on tests/Unit/Domain/FunctionsOfTime/Test_QuaternionFunctionOfTime.cpp:
```diff
@@ -0,0 +1,66 @@
+        expir_time};
+
+    // Update PiecewisePolynomial with 0 2nd derivative so it's constant
+    const double increment = 0.5;
```
maybe rename `increment` -> `delta_t` or `time_step`?

knelli2 commented:
Added PR #3321 to actually check tests.
Added more tests:
1) Test expiration time and time bounds. (As I was writing this, I realized I don't ever actually use the expiration time anywhere. But this FunctionOfTime will definitely expire at some time, but that's determined by the omega PiecewisePolynomial it looks at. Is it worth keeping around or should I just get rid of it if it's not being used?)
2) Constant omega
3) Linear omega using euler angles
4) Quadratic omega using euler angles

knelli2 commented:
An open question I still have is how the `pup`ing is going to work with the stored `omega_f_of_t_ptr_` which points to a PiecewisePolynomial that holds the angular velocity. Because if I store a pointer to it, or even a reference, when the QuaternionFunctionOfTime gets `pup`ed, it could go to a different machine and so the pointer would point to a wrong memory address. But the QuaternionFuntionOfTime still needs to always be able to access the omega PiecewisePolynomial. Unsure how to do this.

knelli2 commented:
Reworked QuaternionFunctionOfTime to now store the omega PiecewisePolynomial internally rather than a pointer to an external one. Because of this, QuaternionFunctionOfTime must be constructed with the initial omega as well as the initial quatenion.

Added an `update(time, updated_max_deriv, next_expr_time)` function that does two things:
  1) First, it updates the stored omega PiecewisePolynomial with the `updated_max_deriv` at the given time
  2) After, the PiecewisePolynomial has been updated, it updates the stored quaternions to be in sync with the updated omegas

Added new member functions. Since we are storing omega which needs to be controlled, I added some unambiguous function calls to get the quaternion or omega at time `t`. They follow the naming convention:
`quat_func(t)`, `quat_func_and_deriv(t)`, `quat_func_and_2_derivs(t)`
`omega_func(t)`, `omega_func_and_deriv(t)`, `omega_func_and_2_derivs(t)`

Additionally, we have the inherited `func` members from FunctionOfTime. I debated what this should return and decided that `QuaternionFunctionOfTime.func(t)` will return omega at time `t` (identical to `omega_func(t)`). The reasoning is this: The control system will control a FunctionOfTime that has an `update` function so that it can reset it periodically (like PiecewisePolynomial). However, to get the function at a specific time, it will call the `func` member of that FunctionOfTime. Since we want the control system as general as possibly, and since we are controlling omega here and not the quaternion (and the `update` function of QuaternionFunctionOfTime updates the stored PiecewisePolynomial), the `QuaternionFunctionOfTime.func(t)` member should return omega at time `t`, not the quaternion. It is slightly confusing having a `QuaternionFunctionOfTime.func(t)` *not* return a quaternion, so I'm open to ideas to improve it. Possibly rename the class?

knelli2 commented:
Edit to above comment:

After discussing with @markscheel and @nilsdeppe, we decided that instead of the control system being as generic as possible, we want the maps to be as generic as possible. This is because we will have rotation maps that use quaternions that don't need to be solved by an ODE (e.g. SettleToConstant). The control system will then be specialized to accommodate a QuaternionFunctionOfTime. Thus, now `QuaternionFunctionOfTime::func(t)` will return the quaternion at time `t`, just like `quat_func(t)`. To access omega, still use `omega_func(t)`. `quat_func(t)` is redundant with `func(t)` but I left it in for symmetry with `omega_func(t)`, but it's very easy to remove if others feel we don't need it.

knelli2 committed:
Allow external read access to stored info in PiecewisePolynomial

knelli2 committed:
Add QuaternionFunctionOfTime

knelli2 commented:
Rebased onto latest develop now that #3321 is merged. No longer dependent (I don't think I have permissions to remove the label).

I also split it into two commits.

wthrowe commented on src/Domain/FunctionsOfTime/CMakeLists.txt:
```diff
@@ -12,6 +12,7 @@ spectre_target_sources(
   FunctionOfTimeHelpers.cpp
   PiecewisePolynomial.cpp
   QuaternionHelpers.cpp
+  QuaternionFunctionOfTime.cpp
```
Alphabet.

wthrowe commented on src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.hpp:
```diff
@@ -0,0 +1,173 @@
+  /// Returns stored omega at an arbitrary time `t`.
+  std::array<DataVector, 1> omega_func(const double t) const noexcept {
+    return omega_f_of_t_.func(t);
+  }
```
Is there a reason to prefer wrappers over exposing a const reference to `omega_f_of_t_`?

knelli2 replied:
Unless there is a case where a user needs access to all the methods of `omega_f_of_t_`, I think a wrapper should suffice. The only time `omega_f_of_t_` is indirectly accessed by the user is during the `update` call to reset the `MaxDeriv`th derivative, and the `omega_func` calls. It is never accessed directly by the user.

wthrowe replied:
OK.

wthrowe commented on src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.hpp:
```diff
@@ -0,0 +1,173 @@
+
+ private:
+  mutable std::vector<FunctionOfTimeHelpers::StoredInfo<1, false>>
+      stored_quaternions_and_times_;
```
This will be in the global cache, so I don't think it is safe to have any mutable data.  I think you just have too many methods marked const.

wthrowe commented on src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.hpp:
```diff
@@ -0,0 +1,173 @@
+  /// quaternion at time `t0` and on output, it stores the result at time `t`
+  void solve_quaternion_ode(
+      gsl::not_null<std::array<DataVector, 1>*> quaternion_to_integrate,
+      double t, double t0) const noexcept;
```
Specifying the integration interval in reverse order seems unintuitive to me.

wthrowe commented on src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.cpp:
```diff
@@ -0,0 +1,233 @@
+QuaternionFunctionOfTime<MaxDeriv>::QuaternionFunctionOfTime(
+    double t, std::array<DataVector, 1> initial_quat_func,
+    std::array<DataVector, MaxDeriv + 1> initial_omega_func,
+    double expiration_time) noexcept
```
const for both doubles.

wthrowe commented on src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.cpp:
```diff
@@ -0,0 +1,233 @@
+    std::array<DataVector, MaxDeriv + 1> initial_omega_func,
+    double expiration_time) noexcept
+    : stored_quaternions_and_times_{{t, std::move(initial_quat_func)}},
+      omega_f_of_t_(t, initial_omega_func, expiration_time),
```
Move `initial_omega_func`.

wthrowe commented on src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.cpp:
```diff
@@ -0,0 +1,233 @@
+void QuaternionFunctionOfTime<MaxDeriv>::update(
+    const double time_of_update, const DataVector& updated_max_deriv,
+    const double next_expiration_time) noexcept {
+  omega_f_of_t_.update(time_of_update, updated_max_deriv, next_expiration_time);
```
Take `updated_max_deriv` by value and move it here.

knelli2 replied:
I had a const reference because `omega_f_of_t_.update` moves `updated_max_deriv` so I didn't want to make an extra copy.

wthrowe replied:
That's why you take by value and move things, actually.  There are two cases: the caller is giving up ownership of the passed in resources (either because the passed thing is a temporary or because it was explicitly moved), or the owner is not.  (These cases have names, but I can never keep them straight.)

| Ownership transfer | update() argument type | origin of update() argument | origin of `omega_f_of_t_.update()` argument |
| ------------------ | ---------------------- | --------------------------- | ------------------------------------------- |
| no                 | `const DataVector&`    | reference                   | copy                                        |
| no                 | `DataVector`           | copy                        | move                                        |
| yes                | `const DataVector&`    | reference                   | copy                                        |
| yes                | `DataVector`           | move                        | move                                        |

So if the caller is not giving up ownership, there is one copy either way.  (The by-value version does incur an extra move, but those are generally assumed to be cheap.)  But, if the caller *is* giving up ownership, taking by value avoids a copy.  (If the argument of `omega_f_of_t_.update()` were a reference, on the other hand, you would get the opposite conclusion and want to take by reference here as well, but that result is more intuitive.)

wthrowe commented on src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.cpp:
```diff
@@ -0,0 +1,233 @@
+  }
+
+  // Will store the solution of the ode
+  std::array<DataVector, 1> quaternion_to_integrate;
```
Why an array?

knelli2 replied:
I was following documentation [here](https://spectre-code.org/namespaceOdeIntegration.html) and didn't realize you didn't have to wrap everything in `std::array`s. I'll change it.

wthrowe commented on src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.cpp:
```diff
@@ -0,0 +1,233 @@
+    normalize_quaternion(make_not_null(&quat));
+
+    stored_quaternions_and_times_.emplace_back(
+        //  FunctionOfTimeHelpers::StoredInfo<1, false>{
```
Remove.

wthrowe commented on src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.cpp:
```diff
@@ -0,0 +1,233 @@
+        const std::array<DataVector, 1> omega = omega_f_of_t_.func(time);
+        dt_state[0] = DataVector{
+            {-0.5 * (state[0][1] * omega[0][0] + state[0][2] * omega[0][1] +
+                     state[0][3] * omega[0][2]),
```
Can you integrate the quaternion directly, instead of the DataVector version?  That should be faster and wouldn't require reimplementing quaternion multiplication.

knelli2 replied:
No. Unfortunately only fundamental types, spectre vectors, and `std::array`s thereof can be used in odeint. Furthermore, the datatype used in odeint has to be iterable, which `boost::math::quaternion` is specifically not iterable.

wthrowe replied:
It works with `boost::numeric::odeint::runge_kutta_dopri5<boost::math::quaternion<double>, double, boost::math::quaternion<double>, double, boost::numeric::odeint::vector_space_algebra>`, although you have to do a slightly sketchy (because it involves no user-defined types) definition of a specialization:
```c++
namespace boost::numeric::odeint {
template <>
struct vector_space_norm_inf<boost::math::quaternion<double>> {
  using result_type = double;
  result_type operator()(const boost::math::quaternion<double>& q) const {
    return sup(q);
  }
};
}  // namespace boost::numeric::odeint                                          
```

Also the errors are a bit different for some reason I don't understand, even without switching to quaternion multiplication.

knelli2 replied:
After reading a lot of documentation, yes I now see that works. Added a new commit that implements this.

wthrowe commented on src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.cpp:
```diff
@@ -0,0 +1,233 @@
+  boost::numeric::odeint::dense_output_runge_kutta<
+      boost::numeric::odeint::controlled_runge_kutta<
+          boost::numeric::odeint::runge_kutta_dopri5<
+              std::array<DataVector, 1>>>>
```
If there's no conversion going on here, make this `auto`.  Seeing the full type isn't informative.

wthrowe commented on src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.cpp:
```diff
@@ -0,0 +1,233 @@
+
+  // Initialize the stepper with initial quaternion (which is the quaternion
+  // passed in), initial time, and initial time step = 1e-3
+  dense_stepper.initialize(*quaternion_to_integrate, t0, 1e-3);
```
How much does the initial time step matter?  Should it scale with something?

knelli2 replied:
The odeint uses adaptive time stepping to stay within error tolerances. Through trial and error, I've found that I get the same result for all time steps less than 1e-3 so I just chose that as a starting value.

wthrowe replied:
That's presumably for some particular problem, though.  Is this number going to work for all the cases we will run into in practice?

knelli2 replied:
This number works for omega constant, linear, and quadratic in the tests. In practice the orbital angular velocity will be almost constant for most of inspiral. Just before merger it *may* be an issue depending on how small the simulation timesteps are. Since it's adaptive, would just setting it to a small value (say 1e-10) suffice, and then the integrator can change the time step however it likes to achieve the appropriate error? Or would that slow things down? (I also changed the current value to 1e-4 in the most recent fixup because changing to integrating quaternions rather than DataVectors somehow changed the error and tests were failing with 1e-3)

wthrowe commented on tests/Unit/Domain/FunctionsOfTime/Test_QuaternionFunctionOfTime.cpp:
```diff
@@ -0,0 +1,247 @@
+
+SPECTRE_TEST_CASE("Unit.Domain.FunctionsOfTime.QuaternionFunctionOfTime",
+                  "[Unit][Domain]") {
+  INFO("QuaternionFunctionOfTime: Expiration time and time bounds") {
```
The `INFO`s should be inside the braces.  They are scoped like variables.

knelli2 committed:
Allow odeint to integrate boost quaternions

wthrowe commented on src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.cpp:
```diff
@@ -0,0 +1,213 @@
+}
+
+template <size_t MaxDeriv>
+std::array<DataVector, 1> QuaternionFunctionOfTime<MaxDeriv>::quat_func(
```
Codecov points out that the functions returning fewer derivatives aren't tested.  Those are easy, so add them.

knelli2 replied:
Is there a way to run codecov locally so I don't have to wait for CI tests to run to know if I missed anything?

wthrowe replied:
I don't know.  @nilsleiffischer?

nilsvu replied:
Yes you can run it locally. Set up a separate build directory with `cmake -D COVERAGE=ON` and run `make unit-test-coverage` (maybe need to `make unit-tests`) before. You'll get the results as HTML output. That said, it might be easier to just push the code and have CI build it, since the separate build with coverage can take up a lot of space and running unit tests with coverage can take a long time (40min on CI).

knelli2 replied:
Thanks!

wthrowe commented on src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.cpp:
```diff
@@ -0,0 +1,213 @@
+  // Integrate from t0 to t, storing result in quaternion_to_integrate
+  boost::numeric::odeint::integrate_adaptive(
+      dense_stepper, quaternion_ode_system, *quaternion_to_integrate, t0, t,
+      1e-4);
```
In it's infinite wisdom, GitHub has apparently decided I can't reply to the old thread, so, new continuation thread.

Perhaps a better way to express my concern here is to point out that this number isn't dimensionless.

Anyway, though, I've thought about this and this value is probably fine for any case we will hit.  The required value probably scales with 1/omega, and omega can't really exceed 1/(total mass).  We always choose the total mass to be 1 or use physical values of >1 solar mass.  Your test uses omega > 1, so that should be more restrictive.

wthrowe reviewed (changes requested):
You can squash the existing fixup commits.

wthrowe reviewed (commented):
Looks good.  Squash.

@markscheel Ready to take another look at this?  It's had rather significant changes since your last review.

knelli2 commented:
Rebased on develop and added a fixup commit. When addressing the couple lines that weren't checked in codecov, I realized that this loop was unnecessary because we update the PiecewisePolynmial immediately before we update the quaternions, so there would only ever be one missing quaternion.

@wthrowe could you take another quick look?

wthrowe commented:
Looks good.  Go ahead and squash.

knelli2 commented:
Codecov doesn't like that I don't test the constructor that takes a `CkMigrateMessage*` which is unused. Is there a way to ignore this or should I just leave it?

nilsdeppe commented:
I'm not aware of any way to test that constructor, so I would say ignore codecov

nilsvu commented:
@knelli2 You can exclude lines from coverage checks with `LCOV_EXCL_*` comments, you'll find examples in the repo if you search for the string

knelli2 commented:
Thank you @nilsleiffischer!
@wthrowe I added the `LCOV_EXCL` comments in and squashed together

markscheel commented on src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.hpp:
```diff
@@ -0,0 +1,175 @@
+      stored_quaternions_and_times_;
+
+  domain::FunctionsOfTime::PiecewisePolynomial<MaxDeriv> omega_f_of_t_;
+  double expiration_time_{std::numeric_limits<double>::lowest()};
```
Unless I am missing something, it seems that `expiration_time_` should always be the same as `omega_f_of_t_.expiration_time_`.  How about eliminating the member variable `expiration_time_` (so you don't need to initialize, reset, or pup it, or worry about bugs where the two expiration times might accidentally be different), and then use `omega_f_of_t_.time_bounds()` to get expiration_time where needed.

(If the two expiration times are allowed to be different, then please add a comment saying why).

knelli2 replied:
Technically they could be different, with the requirement that the domain of validity of the quaternions is contained within the domain of validity of the internal PiecewisePolynomial. However, I don't think we want that because then we'd have to constantly check that the expiration times are correct relative to each other. They should just be the same all the time.

I'll get rid of the member variable and just use the internal PiecewisePolynomial time bounds and expiration time.

markscheel reviewed (commented):
Looks good. One comment about possibly eliminating a member variable.

wthrowe commented:
I approve of the latest change.

nilsvu commented:
@markscheel Please take a look at the fixup

markscheel reviewed (dismissed):
Please squash

knelli2 commented:
Squashed and rebased on latest develop :)
